docker/lib/dependabot/docker/tag.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+module Dependabot
+  module Docker
+    class Tag
+      VERSION_REGEX = /v?(?<version>[0-9]+(?:\.[0-9]+)*(?:_[0-9]+|\.[a-z0-9]+|-(?:kb)?[0-9]+)*)/i
+      VERSION_WITH_SFX = /^#{VERSION_REGEX}(?<suffix>-[a-z][a-z0-9.\-]*)?$/i
+      VERSION_WITH_PFX = /^(?<prefix>[a-z][a-z0-9.\-]*-)?#{VERSION_REGEX}$/i
+      VERSION_WITH_PFX_AND_SFX = /^(?<prefix>[a-z\-]+-)?#{VERSION_REGEX}(?<suffix>-[a-z\-]+)?$/i
+      NAME_WITH_VERSION =
+        /
+          #{VERSION_WITH_PFX}|
+          #{VERSION_WITH_SFX}|
+          #{VERSION_WITH_PFX_AND_SFX}
+      /x
+
+      attr_reader :name
+
+      def initialize(name)
+        @name = name
+      end
+
+      def to_s
+        name
+      end
+
+      def comparable?
+        name.match?(NAME_WITH_VERSION)
+      end
+
+      def same_precision?(other)
+        other.precision == precision
+      end
+
+      def canonical?
+        return false unless numeric_version
+        return true if name == numeric_version
+
+        # .NET tags are suffixed with -sdk
+        return true if name == numeric_version + "-sdk"
+
+        name == "jdk-" + numeric_version
```

Comment:
Definitely weird! The code was already there and I wanted to focus on the refactoring of extracting a `Dependabot::Docker::Tag` class, so I decided to stop there and leave existing logic untouched.

But I can improve this as a follow, since I also just realized that we can remove the `-sdk` special case now since these images [have been removed](https://github.com/dotnet/dotnet-docker/issues/3052) 🎉